### PR TITLE
 Modernize PInvoke with CSWin32

### LINKED ISSUE #12803: Refresh usage of Windows Credential manager to properly store credentials
Current implementation lacks compatibility and is prone to errors due to length limitations of password blobs in Windows Credential Manager.
Review implementation to include "attributes"/pages for increased payload size of up to 64 KiB - should be enough for a password.

### LINKED ISSUE #12804: Modernize PInvoke with CSWin32
Replace all custom PInvoke signatures etc. with properly generated methods from https://github.com/microsoft/cswin32.

### PR DESCRIPTION
**Pull request by AliveDevil**
First part of #12805.

Fixes #12803 
Fixes #12804

### CODE REVIEW

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs [hunk 1]
```diff
@@ -49,30 +48,178 @@ public override void addPassword(Scheme scheme, int port, string hostName, strin
...
+                if (credential.getOauth().getExpiryInMilliseconds() != null)
+                {
+                    winCred.Attributes["OAuth Expiry"] = credential.getOauth().getExpiryInMilliseconds().longValue().ToString();
+                }
+            }
+            if (protocol.isPrivateKeyConfigurable())
+            {
+                winCred.Attributes["Private Key Passphrase"] = credential.getIdentityPassphrase();
+            }
+            if (credential.isPasswordAuthentication())
```
Comment (AliveDevil) [inline core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:184]: Need to check here.
If password authentication is required, then if either username or password is empty the whole credential is not saved.

Can private key configurable be used with password authentication, and username is empty?

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs [hunk 2]
```diff
@@ -49,30 +48,178 @@ public override void addPassword(Scheme scheme, int port, string hostName, strin
...
+            if (protocol.isHostnameConfigurable() || !(protocol.isTokenConfigurable() || protocol.isOAuthConfigurable()))
+            {
+                pathBuilder.Append(":" + bookmark.getHostname());
+                if (protocol.isPortConfigurable() && !Equals(protocol.getDefaultPort(), bookmark.getPort()))
+                {
+                    pathBuilder.Append(":" + bookmark.getPort());
+                }
+            }
+            targetBuilder.Path = pathBuilder.ToString();
+            targetBuilder.Query = "user=" + credentials.getUsername();
```
Comment (AliveDevil) [inline core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:220]: Uses format `application.container.name:protocol.identifier(:port)?\?user=username`.
Should this be changed/updated?

**Comment by AliveDevil** [comment]
Tests won't compile, need to fix that.

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs [hunk 3]
```diff
@@ -40,7 +39,7 @@ public override void addPassword(string serviceName, string user, string passwor
             var hostUrl = $"{serviceName} - {user}";
             if (!WinCredentialManager.SaveCredentials(hostUrl, new NetworkCredential(user, password)))
             {
-                throw new LocalAccessDeniedException($"Could not save credentials for \"{hostUrl}\" to Windows Credential Manager.");
+                logger.error($"Could not save credentials for \"{hostUrl}\" to Windows Credential Manager.");
```
Comment (dkocher) [inline core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:42]: Why no longer throwing failure?

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs [hunk 4]
```diff
@@ -17,15 +17,14 @@
...
-using System.Collections.Generic;
-using System.Linq;
 using System.Net;
 using System.Text;
-using System.Threading.Tasks;
+using static Windows.Win32.Security.Credentials.CRED_PERSIST;
+using static Windows.Win32.Security.Credentials.CRED_TYPE;
 
 namespace Ch.Cyberduck.Core
 {
```
Comment (dkocher) [inline core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:30]: Should probably just implement `HostPasswordStore` if overriding all implementations from `DefaultHostPasswordStore` anyways.

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs (hunk 1 above)
Comment (dkocher) [reply core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:184]: Should match logic in `DefaultHostPasswordStore` to make sure we have no regressions.

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs (hunk 3 above)
Comment (AliveDevil) [reply core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:42]: Is there a difference in throwing? Will be written into error log regardless, and prevents unecessary stack rewinds and exception handling.

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs (hunk 4 above)
Comment (AliveDevil) [reply core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:30]: I'd agree, unless there is logic in DefaultHostPasswordStore, that I don't want to replicate _again_. Thats why I overwrote the HostPasswordStore-implementation for the revised implementation, and falling back to DefaultHostPasswordStore-logic.
See original logic for CredentialManagerPasswordStore: https://github.com/iterate-ch/cyberduck/blob/master/core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs
I didn't want to have three different proxies in https://github.com/iterate-ch/cyberduck/blob/feature/GH-12805-1/core/src/main/csharp/ch/cyberduck/core/PasswordStoreFacade.cs.

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs (hunk 1 above)
Comment (AliveDevil) [reply core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:184]: DefaultHostPasswordStore runs separate addPassword-calls for these conditions.

CredentialManagerPasswordStore now only performs one single commit with all fields combined.

**Comment by AliveDevil** [comment]
Works For Me (tm)
```
[INFO]   [msbuild] C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(437,5): error MSB8013: This project doesn't contain the Configuration and Platform combination of Release|x64. [D:\a\cyberduck\cyberduck\windows\src\main\wix\WindowsVersionCustomAction\WindowsVersionCustomAction.vcxproj]
```

**Comment by tatze96** [comment]
[637825876270682181.txt](https://github.com/iterate-ch/cyberduck/files/8230295/637825876270682181.txt)
Cyberduck crasht beim verwenden des Disconnect Buttons.

**Comment by AliveDevil** [comment]
Wouldn't merge for 8.3.

File: core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs (hunk 2 above)
Comment (AliveDevil) [reply core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:228]: Upping this one.

`application.container.name` for Cyberduck is `duck`, for Mountain Duck Windows this is `Cyberduck`. This way Cyberduck and Mountain Duck don't share credentials, which is kind of unfortunate. Which key should be used for these shared elements? Taking into account branded builds.

**Review by dkocher — APPROVED** [review]

**Review by ylangisc — APPROVED** [review]

**Pull request merged**